x/yarpctest/service.go

```diff
@@ -38,57 +40,86 @@ import (
 
 // HTTPService will create a runnable HTTP service.
 func HTTPService(options ...api.ServiceOption) api.Lifecycle {
-	opts := api.ServiceOpts{}
-	for _, option := range options {
-		option.ApplyService(&opts)
-	}
-	if opts.Listener != nil {
-		if err := opts.Listener.Close(); err != nil {
-			panic(err)
+	return startThatCreatesStopFunc(func(t testing.TB) (stopper func(testing.TB) error, startErr error) {
+		opts := api.ServiceOpts{}
+		for _, option := range options {
+			option.ApplyService(&opts)
 		}
-	}
-	inbound := http.NewTransport().NewInbound(fmt.Sprintf("127.0.0.1:%d", opts.Port))
-	return createService(opts.Name, inbound, opts.Procedures, options)
+		if opts.Listener != nil {
+			require.NoError(t, opts.Listener.Close())
+		}
+		inbound := http.NewTransport().NewInbound(fmt.Sprintf("127.0.0.1:%d", opts.Port))
+		s := createService(opts.Name, inbound, opts.Procedures, options)
+		return s.Stop, s.Start(t)
+	})
 }
 
 // TChannelService will create a runnable TChannel service.
 func TChannelService(options ...api.ServiceOption) api.Lifecycle {
-	opts := api.ServiceOpts{}
-	for _, option := range options {
-		option.ApplyService(&opts)
-	}
-	if opts.Listener != nil {
-		if err := opts.Listener.Close(); err != nil {
-			panic(err)
+	return startThatCreatesStopFunc(func(t testing.TB) (stopper func(testing.TB) error, startErr error) {
+		opts := api.ServiceOpts{}
+		for _, option := range options {
+			option.ApplyService(&opts)
 		}
-	}
-	trans, err := tchannel.NewTransport(
-		tchannel.ListenAddr(fmt.Sprintf("127.0.0.1:%d", opts.Port)),
-		tchannel.ServiceName(opts.Name),
-	)
-	if err != nil {
-		panic(err)
-	}
-	inbound := trans.NewInbound()
-	return createService(opts.Name, inbound, opts.Procedures, options)
+		listener := opts.Listener
+		var err error
+		if listener == nil {
+			listener, err = net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", opts.Port))
+			require.NoError(t, err)
+		}
+		trans, err := tchannel.NewTransport(
+			tchannel.ServiceName(opts.Name),
+			tchannel.Listener(listener),
+		)
+		require.NoError(t, err)
+		inbound := trans.NewInbound()
+		s := createService(opts.Name, inbound, opts.Procedures, options)
+		return s.Stop, s.Start(t)
+	})
 }
 
 // GRPCService will create a runnable GRPC service.
 func GRPCService(options ...api.ServiceOption) api.Lifecycle {
-	opts := api.ServiceOpts{}
-	for _, option := range options {
-		option.ApplyService(&opts)
+	return startThatCreatesStopFunc(func(t testing.TB) (stopper func(testing.TB) error, startErr error) {
+		opts := api.ServiceOpts{}
+		for _, option := range options {
+			option.ApplyService(&opts)
+		}
+		trans := grpc.NewTransport()
+		listener := opts.Listener
+		var err error
+		if listener == nil {
+			listener, err = net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", opts.Port))
+			require.NoError(t, err)
+		}
+		inbound := trans.NewInbound(listener)
+		service := createService(opts.Name, inbound, opts.Procedures, options)
+		return service.Stop, service.Start(t)
```

Comment:
The `startThatCreatesStopFunc` took me a bit to understand what was going on. I think we could simplify this if we embedded the listening logic into a `wrappedDispatcher`. So, `wrappedDispatcher.Start()` could call a `listen() err` method before the other start logic.

eg this could turn the `func *Service()` functions into:
```
listen = someFunc() err{ 
  /* shove listening logic in here */ 
}
return createService (..., listen)
```
Then the `wrappedDispatcher.Start()` could just add:
```
func (w *wrappedDispatcher) Start(t testing.TB) error {
  if err := w.listen(); err != nil{
    return err
  }
  
  /* run whatever is in wrappedDispatcher.Start() right now */
}
```
I'm not sure if this would mess up other parts of the test infrastructure though that may not need a listener.

Reply:
Unfortunately we need the listener before we've even created the dispatcher since we need to pass it into the transport, not a really good way around that unfortunately.